src/main/kotlin/app/revanced/patches/all/telephony/sim/spoof/SpoofSimCountryPatch.kt

```diff
@@ -0,0 +1,76 @@
+package app.revanced.patches.all.telephony.sim.spoof
+
+import app.revanced.patcher.extensions.InstructionExtensions.getInstruction
+import app.revanced.patcher.extensions.InstructionExtensions.removeInstruction
+import app.revanced.patcher.extensions.InstructionExtensions.replaceInstruction
+import app.revanced.patcher.patch.annotation.Patch
+import app.revanced.patcher.patch.options.types.StringPatchOption.Companion.stringPatchOption
+import app.revanced.patcher.util.proxy.mutableTypes.MutableMethod
+import app.revanced.util.patch.AbstractTransformInstructionsPatch
+import com.android.tools.smali.dexlib2.builder.instruction.BuilderInstruction11x
+import com.android.tools.smali.dexlib2.iface.ClassDef
+import com.android.tools.smali.dexlib2.iface.Method
+import com.android.tools.smali.dexlib2.iface.instruction.Instruction
+import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction
+import com.android.tools.smali.dexlib2.iface.reference.MethodReference
+import java.util.*
+
+
+@Patch(
+    name = "Spoof SIM country",
+    description = "Spoofs country information returned by the SIM card provider.",
+    use = false,
+)
+@Suppress("unused")
+object SpoofSimCountryPatch : AbstractTransformInstructionsPatch<Pair<Int, String>>() {
+    private val networkCountryIso by stringPatchOption(
+        "getNetworkCountryIso",
+        null,
+        "Network ISO Country Code",
+        "ISO-3166-1 alpha-2 country code equivalent of the MCC (Mobile Country Code) " +
+                "of the current registered operator or the cell nearby.",
+        validator = { it?.uppercase() in Locale.getISOCountries() || it == null }
+    )
+
+    private val simCountryIso by stringPatchOption(
+        "getSimCountryIso",
+        null,
+        "Sim ISO Country Code",
+        "ISO-3166-1 alpha-2 country code equivalent for the SIM provider's country code.",
+        validator = { it?.uppercase() in Locale.getISOCountries() || it == null }
+    )
+
+    override fun filterMap(
+        classDef: ClassDef,
+        method: Method,
+        instruction: Instruction,
+        instructionIndex: Int
+    ): Pair<Int, String>? {
+        if (instruction !is ReferenceInstruction) return null
+
+        val reference = instruction.reference as? MethodReference ?: return null
+        if (reference.definingClass != "Landroid/telephony/TelephonyManager;") return null
+
+        if (!options.contains(reference.name)) return null
+        return (options[reference.name].value as? String)?.let { instructionIndex to it.trim() }
```

Comment:
This is not ideal. You should not use the option key for this purpose. Create an enum; if the reference matches, then you get the replacement value from the option.

Reply:
I need input on this, how else would you do it:

1.
```
    private val networkCountryIso by stringPatchOption()

    private val simCountryIso by stringPatchOption()

    override fun filterMap(
        /* ... */
    ): Pair<Int, String>? {
        /* ... */

        val methodMatch = MethodCall.entries.firstOrNull {
            /* ... */
        }

        val optionValue = options[methodMatch.optionKey].value as? String ?: return null
        return instructionIndex to optionValue.lowercase()
    }
 
    private enum class MethodCall(
        /* ... */
        val optionKey: String
    ) {
        NetworkCountryIso(
            "networkCountryIso"
        ),
        SimCountryIso(
            "simCountryIso"
        )
    }

```

2.
```
    private val networkCountryIso by stringPatchOption()

    private val simCountryIso by stringPatchOption()

    override fun filterMap(
        /* ... */
    ): Pair<Int, String>? {
        /* ... */

        val methodMatch = MethodCall.entries.firstOrNull {
            /* ... */
        }

        return methodMatch.optionValue()?.let { instructionIndex to it.lowercase() }
    }
 
    // Lambda because we don't want the previous option value to persist if the patch is cancelled mid-execution
    private enum class MethodCall(
        /* ... */
        val optionValue: () -> String?
    ) {
        NetworkCountryIso(
            optionValue = { networkCountryIso }
        ),
        SimCountryIso(
            optionValue = { simCountryIso }
        )
    }

```

Reply:
`MethodCall` should be an enum of methods you want to transform and not contain the options key. See 
https://github.com/revanced/revanced-patches/blob/aeb5299ca50cdeef1c6db273d312b2afea82af7b/src/main/kotlin/app/revanced/patches/all/connectivity/wifi/spoof/SpoofWifiPatch.kt#L44-L47

Then, you can check if the current instruction matches the MethodCall enum and replace it with the value of the respective option.

Reply:
```kotlin

        val optionValue = when(methodMatch) {
            MethodCall.NetworkCountryIso -> networkCountryIso
            MethodCall.SimCountryIso -> simCountryIso
        }
        return optionValue?.let { instructionIndex to it.lowercase() }

``` 

Is this fine?

Reply:
Yeah, works:

```kotlin

        val iso = when(methodMatch) {
            MethodCall.NetworkCountryIso -> networkCountryIso
            MethodCall.SimCountryIso -> simCountryIso
        }.toLowerCase()

        return optionValue?.let { instructionIndex to iso }

```

Reply:
Thank you for the suggestions.

Reply:
Whoops the null check on optionValue may not be correct